libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -513,26 +517,39 @@ method removeValidator*(p: PubSub,
 
 method validate*(p: PubSub, message: Message): Future[ValidationResult] {.async, base.} =
   var pending: seq[Future[ValidationResult]]
+  result = ValidationResult.Accept
+
   trace "about to validate message"
   for topic in message.topicIds:
     trace "looking for validators on topic", topicId = topic,
                                              registered = toSeq(p.validators.keys)
     if topic in p.validators:
       trace "running validators for topic", topicId = topic
       for validator in p.validators[topic]:
-        pending.add(validator(topic, message))
-
-  result = ValidationResult.Accept
-  let futs = await allFinished(pending)
-  for fut in futs:
-    if fut.failed:
-      result = ValidationResult.Reject
-      break
-    let res = fut.read()
-    if res != ValidationResult.Accept:
-      result = res
-      if res == ValidationResult.Reject:
+        case p.validationStrategy
+          of ValidationStrategy.Paralel:
+            pending.add(validator(topic, message))
+          of ValidationStrategy.Sequential:
+            let validatorRes = await validator(topic, message)
+            # early break on first Reject/Ignore
+            if validatorRes != ValidationResult.Accept:
+              result = validatorRes
+              # TODO: Wrong, actually need to break two loops
+              # but don't get why message.topicIds is a seq. This
+              # will just work with message.topicIds.len = 1
```

Comment:
@Menduist Not sure I understand why message.topicIds is a seq. Guess a message can only come from one topic?

But apart from that, any idea on how to break these two loops in here? In python you have the `for ... else` but unsure in nim.

```
for i in l1:
    for j in l2:
        print(i, j)
        if i == 2 and j == 20:
            break
    else:
        continue
    break
```

Reply:
The specs allow multiple topics per message, but it's a minefield.

You can use a named block that you can break

Reply:
ah right, done c1fe209b3dcc10aff362c9ae3a218f5622aaaf5f